javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -164,54 +164,53 @@ private boolean arePointingToSameLeaf(int i, int j) {
     }
 
     /* descend the tree from root to leaf, applying the given modifications along the way, returning the new root */
-    private Object[] modifyLeaf(Object[] root, int depthShift, int index, NodeModifier node, NodeModifier leaf) {
+    private Object modifyLeaf(Object root, int depthShift, int index, NodeModifier node, NodeModifier leaf) {
         if (depthShift == 0) {
             return leaf.apply(root, index);
         } else {
             int previousIndex = firstDigit(index, depthShift);
             root = node.apply(root, previousIndex);
 
-            Object[] array = root;
+            Object array = root;
             for (int shift = depthShift - BRANCHING_BASE; shift >= BRANCHING_BASE; shift -= BRANCHING_BASE) {
                 final int offset = digit(index, shift);
 
-                final Object previous = array[previousIndex];
-                final Object[] newNode = node.apply(previous, offset);
-                array[previousIndex] = newNode;
+                final Object previous = getAt(array, previousIndex);
+                final Object newNode = node.apply(previous, offset);
+                setAt(array, previousIndex, newNode);
```

Comment:
since we have `Object`s everywhere, array notation cannot be used anymore, since the `Object[]` type cannot refer to primitive arrays, but `Object` can